test/functional/p2p_invalid_messages.py

```diff
@@ -77,6 +79,13 @@ def test_buffer(self):
         conn.sync_with_ping(timeout=1)
         self.nodes[0].disconnect_p2ps()
 
+    def test_duplicate_version_msg(self):
+        self.log.info("Test duplicate version message is ignored")
+        conn = self.nodes[0].add_p2p_connection(P2PDataStore())
+        with self.nodes[0].assert_debug_log(['redundant version message from peer']):
+            conn.send_and_ping(msg_version())
```

Comment:
I think it's much better to test the behaviour rather than the log. Here, that would be sending the message and then verifying that the peer is still connected (you could even check that `getpeerinfo.banscore = 0`, but that also seems a bit too implementation specific).

Reply:
The `_and_ping` part will take care of asserting that the connection is still up, no?

Reply:
Yes, I think that would do it, which makes asserting on the log redundant.

Reply:
I like the assert log for two reasons:
* I can call `git grep -W 'redundant version message'` and see the code as well as the test without searching manually
* It protects against test mistakes, such as sending a `verion` msg, which would also be ignored

Reply:
ok